### PR TITLE
CSP-1369 your training providers page

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/SFA.DAS.Employer.PR.Domain.UnitTests/SFA.DAS.Employer.PR.Domain.UnitTests.csproj
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Domain.UnitTests/SFA.DAS.Employer.PR.Domain.UnitTests.csproj:1]: Stop removing the last line, is this done deliberately or your VS is doing it?

File: src/SFA.DAS.Employer.PR.Domain/Models/AccountLegalEntity.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+public class AccountLegalEntity
+{
+    public long Id { get; set; }
+    public string PublicHashedId { get; set; }
+
+    public string Name { get; set; }
+    public long AccountId { get; set; }
+    public List<Permission> Permissions { get; set; } = new();
+}
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Domain/Models/AccountLegalEntity.cs]: Add new line at the end

File: src/SFA.DAS.Employer.PR.Domain/Models/EmployerIdentifier.cs [hunk 2]
```diff
@@ -1,4 +1,21 @@
-﻿namespace SFA.DAS.Employer.PR.Domain.Models;
+﻿using Newtonsoft.Json;
 
-public record EmployerIdentifier(string AccountId, string EmployerName, string Role);
+namespace SFA.DAS.Employer.PR.Domain.Models;
+
+public class EmployerIdentifier
+{
+    [JsonProperty("encodedAccountId")]
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Domain/Models/EmployerIdentifier.cs]: Can we avoid using Newtonsoft? The deserializer is using System.Text.Json anyway

File: src/SFA.DAS.Employer.PR.Domain/Models/AccountLegalEntity.cs
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Domain/Models/AccountLegalEntity.cs:1]: Are these models generic, if not they should be part of the handler

File: src/SFA.DAS.Employer.PR.Domain/OuterApi/Responses/GetEmployerUserAccountsResponse.cs [hunk 3]
```diff
@@ -1,5 +1,4 @@
 ﻿namespace SFA.DAS.Employer.PR.Domain.OuterApi.Responses;
 public record GetEmployerUserAccountsResponse(string FirstName, string LastName, string EmployerUserId, bool IsSuspended, IEnumerable<EmployerUserAccountItem> UserAccounts);
 
-public record EmployerUserAccountItem(string EncodedAccountId, string DasAccountName, string Role);
-
+public record EmployerUserAccountItem(string EncodedAccountId, string DasAccountName, string Role);
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Domain/OuterApi/Responses/GetEmployerUserAccountsResponse.cs]: I see a lot of these where the last new line has been removed. Github shows missing new line at the end, so I would like to keep these newlines.

File: src/SFA.DAS.Employer.PR.Web.UnitTests/AppStart/WhenHandlingEmployerAllRolesAuthorizationHandler.cs
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web.UnitTests/AppStart/WhenHandlingEmployerAllRolesAuthorizationHandler.cs:1]: We do not follow these kind of naming conventions, can you follow our standard naming convention please

File: src/SFA.DAS.Employer.PR.Web.UnitTests/AppStart/WhenHandlingEmployerAccountAuthorization.cs
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web.UnitTests/AppStart/WhenHandlingEmployerAccountAuthorization.cs:1]: naming convention is incorrect

File: src/SFA.DAS.Employer.PR.Web/Models/PermissionModel.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    public const string PermissionToAddRecordsText = "Yes, employer will review records";
+    public const string PermissionToRecruitReviewAdvertsText = "Yes, employer will review adverts";
+    public const string PermissionToRecruitText = "Yes";
+    public const string NoPermissionToAddRecordsText = "No";
+    public const string NoPermissionToRecruitText = "No";
+    public long Ukprn { get; set; }
+    public string ChangePermissionsLink { get; set; } = null!;
+    public string ProviderName { get; set; } = null!;
+    public string PermissionToAddRecords { get; set; } = null!;
+    public string PermissionToRecruitApprentices { get; set; } = null!;
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web/Models/PermissionModel.cs:16]: These should be nullable as a value is only assigned if permissions are given.

File: src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAccountAuthorizationHandler.cs [hunk 5]
```diff
@@ -36,7 +36,7 @@ protected override Task HandleRequirementAsync(AuthorizationHandlerContext conte
         return Task.CompletedTask;
     }
 
-    private bool IsEmployerAuthorised(AuthorizationHandlerContext context)
+    public bool IsEmployerAuthorised(AuthorizationHandlerContext context)
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAccountAuthorizationHandler.cs]: Why public?

File: src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAccountAuthorizationHandler.cs [hunk 6]
```diff
@@ -1,11 +1,11 @@
...
-using System.Security.Claims;
-using Microsoft.AspNetCore.Authorization;
+﻿using Microsoft.AspNetCore.Authorization;
 using Newtonsoft.Json;
 using SFA.DAS.Employer.PR.Domain.Interfaces;
 using SFA.DAS.Employer.PR.Domain.OuterApi.Responses;
 using SFA.DAS.Employer.PR.Web.Infrastructure;
+using System.Diagnostics.CodeAnalysis;
+using System.IdentityModel.Tokens.Jwt;
+using System.Security.Claims;
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAccountAuthorizationHandler.cs:7]: Visual studio settings are not correct, system first.

File: src/SFA.DAS.Employer.PR.Web/Infrastructure/Services/UserRolesService.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+public class UserRolesService : IUserRolesService
+{
+    private readonly ILogger<UserRolesService> _logger;
+
+    public UserRolesService(ILogger<UserRolesService> logger)
+    {
+        _logger = logger;
+    }
+
+    public bool CheckUserAccountAccess(ClaimsPrincipal user, string employerAccountId, EmployerUserRole userRoleRequired)
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web/Infrastructure/Services/UserRolesService.cs]: This can be extension to `ClaimsPrincipal` Also can we avoid using Newtonsoft?

File: src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAllRolesAuthorizationHandler.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
+﻿using Microsoft.AspNetCore.Authorization;
+using Newtonsoft.Json;
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAllRolesAuthorizationHandler.cs]: Can we avoid using Newstonsoft?

File: src/SFA.DAS.Employer.PR.Web.UnitTests/Controllers/YourTrainingProvidersControllerTests.cs [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+using SFA.DAS.Employer.PR.Web.Infrastructure.Services;
+using SFA.DAS.Employer.PR.Web.Models;
+using SFA.DAS.Testing.AutoFixture;
+
+namespace SFA.DAS.Employer.PR.Web.UnitTests.Controllers;
+public class YourTrainingProvidersControllerTests
+{
+
+    [Test, MoqAutoData]
+    public void GetAccountLegalEntities_ReturnsApiResponses(
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web.UnitTests/Controllers/YourTrainingProvidersControllerTests.cs]: ApiResponse?

File: src/SFA.DAS.Employer.PR.Domain/Models/AccountLegalEntity.cs
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Domain/Models/AccountLegalEntity.cs:1]: I was in 2 minds about this one, I can reuse a fair bit of it in my next story.  I will redo so that the parts I reuse are in the model, and the specific elements for the handler can inherit it and extend it.

File: src/SFA.DAS.Employer.PR.Domain.UnitTests/SFA.DAS.Employer.PR.Domain.UnitTests.csproj
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Domain.UnitTests/SFA.DAS.Employer.PR.Domain.UnitTests.csproj:1]: VS is doing it.  I'll try to reset

File: src/SFA.DAS.Employer.PR.Domain/Models/AccountLegalEntity.cs
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Domain/Models/AccountLegalEntity.cs:1]: This model is returned from the outer api and call will be used in multiple stories, so this is the correct place

File: src/SFA.DAS.Employer.PR.Domain/Models/AccountLegalEntity.cs (hunk 1 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Domain/Models/AccountLegalEntity.cs]: ok

File: src/SFA.DAS.Employer.PR.Domain/Models/EmployerIdentifier.cs (hunk 2 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Domain/Models/EmployerIdentifier.cs]: ok

File: src/SFA.DAS.Employer.PR.Web.UnitTests/AppStart/WhenHandlingEmployerAccountAuthorization.cs
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web.UnitTests/AppStart/WhenHandlingEmployerAccountAuthorization.cs:1]: ok

File: src/SFA.DAS.Employer.PR.Web.UnitTests/AppStart/WhenHandlingEmployerAllRolesAuthorizationHandler.cs
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web.UnitTests/AppStart/WhenHandlingEmployerAllRolesAuthorizationHandler.cs:1]: ok

File: src/SFA.DAS.Employer.PR.Web.UnitTests/Controllers/YourTrainingProvidersControllerTests.cs (hunk 9 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web.UnitTests/Controllers/YourTrainingProvidersControllerTests.cs]: renamed

File: src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAccountAuthorizationHandler.cs (hunk 5 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAccountAuthorizationHandler.cs]: I got stuck trying to figure out how to test this when it was private.  I figured it out now.

File: src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAllRolesAuthorizationHandler.cs (hunk 8 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAllRolesAuthorizationHandler.cs]: yes,I've removed all references

File: src/SFA.DAS.Employer.PR.Web/Models/PermissionModel.cs (hunk 4 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web/Models/PermissionModel.cs:16]: No they are always given a value.  If no permissions given, they are set to 'No'

File: src/SFA.DAS.Employer.PR.Web/Infrastructure/Services/UserRolesService.cs (hunk 7 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web/Infrastructure/Services/UserRolesService.cs]: I've removed newtonsoft references, and I like the idea of moving this to claimsprincipal, so will do

File: src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAccountAuthorizationHandler.cs (hunk 6 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAccountAuthorizationHandler.cs:7]: ok

File: src/SFA.DAS.Employer.PR.Domain/Models/EmployerIdentifier.cs [hunk 10]
```diff
@@ -1,4 +1,16 @@
 ﻿namespace SFA.DAS.Employer.PR.Domain.Models;
 
-public record EmployerIdentifier(string AccountId, string EmployerName, string Role);
+public class EmployerIdentifier
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Domain/Models/EmployerIdentifier.cs]: Just curious why was this changed?

File: src/SFA.DAS.Employer.PR.Web/Extensions/ClaimsPrincipalExtensions.cs [hunk 11]
```diff
@@ -22,6 +23,50 @@ private static Guid GetClaimValue(ClaimsPrincipal principal, string claimType)
...
 
+    public static bool CheckUserAccountAccess(this ClaimsPrincipal user, string employerAccountId, EmployerUserRole userRoleRequired)
+    {
+        var employerAccountClaim = user.FindFirst(c => c.Type.Equals(EmployerClaims.AccountsClaimsTypeIdentifier));
+
+        if (employerAccountClaim?.Value == null)
+            return false;
+
+        EmployerUserAccountItem? employerAccount = null;
+        try
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web/Extensions/ClaimsPrincipalExtensions.cs]: What is the point of catching exception, just let it bubble up, as the expectations is to have serialised content in there.

File: src/SFA.DAS.Employer.PR.Web/Extensions/ClaimsPrincipalExtensions.cs [hunk 12]
```diff
@@ -22,6 +23,50 @@ private static Guid GetClaimValue(ClaimsPrincipal principal, string claimType)
...
+        if (!Enum.TryParse<EmployerUserRole>(employerAccount.Role, true, out var claimUserRole))
+        {
+            return false;
+        }
+
+        switch (userRoleRequired)
+        {
+            case EmployerUserRole.Owner when claimUserRole == EmployerUserRole.Owner:
+            case EmployerUserRole.Transactor when claimUserRole is EmployerUserRole.Owner or EmployerUserRole.Transactor:
+            case EmployerUserRole.Viewer when claimUserRole is EmployerUserRole.Owner or EmployerUserRole.Transactor or EmployerUserRole.Viewer:
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web/Extensions/ClaimsPrincipalExtensions.cs]: Viewer cannot make edits, so surely we want to return false here? In which case the method name needs to reflect this `User.IsOwnerOrTransactor()` seems more appropriate

File: src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAllRolesAuthorizationHandler.cs [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        if (!routeValues!.ContainsKey(RouteValueKeys.EncodedAccountId))
+        {
+            return false;
+        }
+
+        var accountIdFromUrl = routeValues[RouteValueKeys.EncodedAccountId]!.ToString()!.ToUpper();
+        var employerAccountClaim = context.User.FindFirst(c => c.Type.Equals(EmployerClaims.AccountsClaimsTypeIdentifier));
+
+        if (employerAccountClaim?.Value == null)
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAllRolesAuthorizationHandler.cs]: Keep this in one line or use blocks

File: src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAllRolesAuthorizationHandler.cs [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        EmployerUserAccountItem? employerAccount = null;
+        try
+        {
+            var employerAccounts = JsonSerializer.Deserialize<Dictionary<string, EmployerUserAccountItem>>(employerAccountClaim.Value);
+            if (employerAccounts != null && employerAccounts.TryGetValue(accountIdFromUrl, out var value))
+            {
+                employerAccount = value;
+            }
+        }
+        catch
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAllRolesAuthorizationHandler.cs]: I do not like hiding exceptions, if this is unexpected situation, then letting it bubble will highlight the issue, else we will never know why user was not allowed.

File: src/SFA.DAS.Employer.PR.Domain/Models/EmployerIdentifier.cs (hunk 10 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Domain/Models/EmployerIdentifier.cs]: no reason really except to make tests easier to write, but I've reverted this and amended the tests

File: src/SFA.DAS.Employer.PR.Web/Extensions/ClaimsPrincipalExtensions.cs (hunk 11 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web/Extensions/ClaimsPrincipalExtensions.cs]: removed the try catch

File: src/SFA.DAS.Employer.PR.Web/Extensions/ClaimsPrincipalExtensions.cs (hunk 12 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web/Extensions/ClaimsPrincipalExtensions.cs]: Yes, that was a mistake.  I have changed along the lines you said, and changed the method name

File: src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAllRolesAuthorizationHandler.cs (hunk 13 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAllRolesAuthorizationHandler.cs]: ok

File: src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAllRolesAuthorizationHandler.cs (hunk 14 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web/Authentication/EmployerAllRolesAuthorizationHandler.cs]: try catch removed

File: src/SFA.DAS.Employer.PR.Web/Extensions/ClaimsPrincipalExtensions.cs (hunk 12 above)
Comment (MarkFCain) [reply src/SFA.DAS.Employer.PR.Web/Extensions/ClaimsPrincipalExtensions.cs]: As per follow up comment, this is only doing IsOwner() now

File: src/SFA.DAS.Employer.PR.Web.UnitTests/Controllers/YourTrainingProvidersControllerTests.cs [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+namespace SFA.DAS.Employer.PR.Web.UnitTests.Controllers;
+public class YourTrainingProvidersControllerTests
+{
+    [Test, MoqAutoData]
+    public void GetAccountLegalEntities_CallsOuterApiEndpoint(
+     [Frozen] Mock<IOuterApiClient> outerApiMock,
+     GetEmployerRelationshipsQueryResponse response,
+     string employerAccountId
+     )
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web.UnitTests/Controllers/YourTrainingProvidersControllerTests.cs:20]: alignment

**Review by DevenKShah — DISMISSED** [review]

File: src/SFA.DAS.Employer.PR.Web/Views/YourTrainingProviders/Index.cshtml [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+                        <li>review and edit anything a provider does on your behalf</li>
+                    </ol>
+                    <p class="govuk-body">
+                        Your provider will not be able to access your account.
+                    </p>
+                </div>
+            </details>
+            <div class="govuk-button-group">
+                <button type="submit" id="continue" class="govuk-button" data-disable-on-submit="true">Add a training provider</button>
+            </div>
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.PR.Web/Views/YourTrainingProviders/Index.cshtml]: Can this be indented please

**Review by DevenKShah — DISMISSED** [review]

**Review by DevenKShah — APPROVED** [review]
